src/datatablerow/utils.jl

```diff
@@ -0,0 +1,194 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    "source data table"
+    dt::T
+    "number of groups"
+    ngroups::Int
+    "row hashes"
+    rhashes::Vector{UInt}
+    "hashindex -> index of group-representative row"
+    gslots::Vector{Int}
+    "group index for each row"
+    groups::Vector{Int}
+    "permutation of row indices that sorts them by groups"
+    rperm::Vector{Int}
+    "starts of ranges in rperm for each group"
+    starts::Vector{Int}
+    "stops of ranges in rperm for each group"
+    stops::Vector{Int}
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+if !isdefined(Base, :unsafe_get)
+    unsafe_get(x::Nullable) = x.value
+    unsafe_get(x::Any) = x
+end
+
+function hashrows_col!{T<:Nullable}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
```

Comment:
Yes, `hash_colel` can still be specialized, but `hashrows_col!` doesn't need to AFAICT.